File: src/YowThi.Erp.Web/src/features/finance/financeSettlementOptions.ts
import { getApiJson } from '../../app/api/apiTransport';
import type { OperationalLocale } from '../../app/i18n/locale';

export interface FinancePayableSettlementOption {
  payableId: string;
  payableKind: string;
  sourceDisplayName: string;
  outstandingThb: number;
  outstandingVersion: number;
  updatedAt: string;
}

export interface FinanceReceivableSettlementOption {
  receivableId: string;
  salesId: string;
  salesDate: string;
  customerDisplayName: string;
  outstandingThb: number;
  outstandingVersion: number;
  updatedAt: string;
}

interface OptionPage<T> {
  items: T[];
  nextCursor: string | null;
}

interface SettlementOptionsQuery {
  locale: OperationalLocale;
  search?: string;
  cursor?: string | null;
  limit?: number;
  signal?: AbortSignal;
}

export function listFinancePayableOptions(
  query: SettlementOptionsQuery,
): Promise<OptionPage<FinancePayableSettlementOption>> {
  return getSettlementOptions<FinancePayableSettlementOption>('payables', query);
}

export function listFinanceReceivableOptions(
  query: SettlementOptionsQuery,
): Promise<OptionPage<FinanceReceivableSettlementOption>> {
  return getSettlementOptions<FinanceReceivableSettlementOption>('receivables', query);
}

function getSettlementOptions<T>(
  resource: 'payables' | 'receivables',
  query: SettlementOptionsQuery,
): Promise<OptionPage<T>> {
  const parameters = new URLSearchParams();
  const search = query.search?.trim();
  if (search) parameters.set('search', search);
  if (query.cursor) parameters.set('cursor', query.cursor);
  if (query.limit !== undefined) parameters.set('limit', String(query.limit));
  const suffix = parameters.size === 0 ? '' : `?${parameters.toString()}`;

  return getApiJson<OptionPage<T>>(
    `/api/v1/finance/settlement-options/${resource}${suffix}`,
    { locale: query.locale, signal: query.signal },
  );
}

File: src/YowThi.Erp.Application/Finance/FinanceSettlementOptions.cs
namespace YowThi.Erp.Application.Finance;

public sealed record FinanceSettlementOptionsQuery(
    string Locale,
    string? Search,
    int Offset,
    int Limit);

public sealed record FinanceSettlementOptionPage<T>(
    IReadOnlyList<T> Items,
    int? NextOffset);

public sealed record FinancePayableSettlementOption(
    Guid PayableId,
    string PayableKind,
    string SourceDisplayName,
    long OutstandingThb,
    long OutstandingVersion,
    DateTimeOffset UpdatedAt);

public sealed record FinanceReceivableSettlementOption(
    Guid ReceivableId,
    Guid SalesId,
    DateOnly SalesDate,
    string CustomerDisplayName,
    long OutstandingThb,
    long OutstandingVersion,
    DateTimeOffset UpdatedAt);

public interface IFinanceSettlementOptionsReader
{
    ValueTask<FinanceSettlementOptionPage<FinancePayableSettlementOption>> GetPayablesAsync(
        FinanceSettlementOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<FinanceSettlementOptionPage<FinanceReceivableSettlementOption>> GetReceivablesAsync(
        FinanceSettlementOptionsQuery query,
        CancellationToken cancellationToken);
}
